File: sdk/ruby/lib/cordum/cap/codec.rb
# frozen_string_literal: true

module Cordum
  module Cap
    module Codec
      # Serializes a protobuf message deterministically.
      # Ruby's google-protobuf produces deterministic output for proto3 messages.
      def self.marshal_deterministic(msg)
        msg.to_proto
      end

      # Serializes a BusPacket with the signature field cleared,
      # producing the canonical bytes used for signing and verification.
      def self.marshal_unsigned_for_signature(packet)
        clone = packet.class.decode(packet.to_proto)
        clone.signature = ''.b
        marshal_deterministic(clone)
      end
    end
  end
end

File: sdk/dotnet/src/Cordum.Cap/Codec.cs
using Google.Protobuf;
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Deterministic protobuf serialization utilities.
/// </summary>
public static class Codec
{
    /// <summary>
    /// Serialize a protobuf message deterministically.
    /// Map fields are serialized in sorted key order when deterministic mode is enabled.
    /// </summary>
    public static byte[] MarshalDeterministic(IMessage msg)
    {
        var size = msg.CalculateSize();
        var buffer = new byte[size];
        var output = new CodedOutputStream(buffer);
        output.Deterministic = true;
        msg.WriteTo(output);
        output.Flush();
        return buffer;
    }

    /// <summary>
    /// Serialize a BusPacket with the signature field cleared.
    /// Used as the input to ECDSA signing/verification.
    /// </summary>
    public static byte[] MarshalUnsignedForSignature(BusPacket packet)
    {
        var clone = packet.Clone();
        clone.Signature = ByteString.Empty;
        return MarshalDeterministic(clone);
    }
}
